### ISSUE TITLE
resulting Installer needs an --update-existing option

### ISSUE THREAD

**Issue by srossross** [conda/constructor#46]
We need a robust way to update existing installs with a new installer file.

**Comment by ilanschnell** [comment]
This feature already exists.  This is what the `-f` option of the resulting installer is for.

**Comment by csoja** [comment]
I don't think -f is the right approach for an upgrade - even if it does what is intended, I think a -f might make someone think something is wrong that they need to force the install.  What about adding a --upgrade option?  @srossross are there other abilities needed for an upgrade feature that the -f feature does not provide?

**Comment by dsludwig** [comment]
As far as I can tell, `-f` does not support unlinking the old packages that are installed in a prefix. Ideally this feature would act more like a `conda upgrade` where the installer provides a `file://` channel to pull from - allowing a compatible set of packages to be pulled from the new installer, along with all the correct upgrade semantics.

**Comment by ilanschnell** [comment]
It is not exactly clear to me what the update (or upgrade) option should do.  Obviously, it is the responsibility of the creator of an installer that all packages contained in the installer are compatible with each other.  Therefore, updating an existing installation is basically replacing an old installation with the new one.  This is different from the exsiting `-f` option, which merely ignores an existing install prefix and "force" installs all packages on top of what is already there.
I've implemented the new `-u` option in https://github.com/conda/constructor/pull/48

**Comment by csoja** [comment]
what is the next step here? @dsludwig  can you test the open PR and see if it meets your needs?

**Comment by ilanschnell** [comment]
Exactly, this is why a created the branch (with PR), so that @dsludwig can try it out.

**Comment by ilanschnell** [comment]
I've update #48 to not use `rm -rf` on the entire install prefix anymore.  Instead, only the excess conda-meta entries are removed upon update, which is also more consistent with what the Anaconda (and Miniconda) installers are currently doing.

**Comment by csoja** [comment]
@dsludwig can you test the latest?

**Comment by dsludwig** [comment]
Thanks Ilan, I will try this out.

**Comment by dsludwig** [comment]
@ilanschnell, I got this error when testing on OS X (using the Maxiconda example):

```
installing: python.app-1.2-py35_4 ...
ln: ./lib: File exists
Error: post-link failed for: python.app-1.2-py35_4
```

**Comment by ilanschnell** [comment]
@dsludwig I was able to reproduce the problem, and also able to fix it (by adding a missing to call to the pre-unlink script).  The Maxiconda example is working now.

**Comment by dsludwig** [comment]
It looks like some packages aren't cleaned up correctly (see `matplotlib`, `numpy`, `pandas`, `pyparsing`, `scipy`):

```
# packages in environment at /Users/dludwig/code/constructor/examples/maxiconda/env:
#
appnope                   0.1.0                    py35_0    defaults
conda                     4.2.11                   py35_0    defaults
conda-env                 2.6.0                         0    defaults
cycler                    0.10.0                   py35_0    defaults
decorator                 4.0.10                   py35_0    defaults
entrypoints               0.2.2                    py35_0    defaults
freetype                  2.5.5                         1    defaults
icu                       54.1                          0    defaults
ipykernel                 4.5.0                    py35_0    defaults
ipython                   5.1.0                    py35_0    defaults
ipython_genutils          0.1.0                    py35_0    defaults
jinja2                    2.8                      py35_1    defaults
jsonschema                2.5.1                    py35_0    defaults
jupyter_client            4.4.0                    py35_0    defaults
jupyter_core              4.2.0                    py35_0    defaults
libpng                    1.6.22                        0    defaults
lighttpd                  1.4.39                        0    defaults
markupsafe                0.23                     py35_2    defaults
matplotlib                1.4.3                     <pip>
matplotlib                1.5.3               np111py35_1    defaults
mistune                   0.7.3                    py35_1    defaults
mkl                       11.3.3                        0    defaults
nbconvert                 4.2.0                    py35_0    defaults
nbformat                  4.1.0                    py35_0    defaults
nomkl                     1.0                           0    defaults
notebook                  4.2.3                    py35_0    defaults
numpy                     1.11.1             py35_nomkl_0  [nomkl]  defaults
numpy                     1.9.3                     <pip>
openssl                   1.0.2j                        0    defaults
pandas                    0.19.0              np111py35_0    defaults
pandas                    0.16.2                    <pip>
path.py                   8.2.1                    py35_0    defaults
pexpect                   4.0.1                    py35_0    defaults
pickleshare               0.7.4                    py35_0    defaults
prompt_toolkit            1.0.8                    py35_0    defaults
ptyprocess                0.5.1                    py35_0    defaults
pycosat                   0.6.1                    py35_1    defaults
pygments                  2.1.3                    py35_0    defaults
pyparsing                 2.0.3                     <pip>
pyparsing                 2.1.4                    py35_0    defaults
python                    3.5.2                         0    defaults
python-dateutil           2.5.3                    py35_0    defaults
python.app                1.2                      py35_4    defaults
pytz                      2016.7                   py35_0    defaults
pyzmq                     16.0.0                   py35_0    defaults
readline                  6.2                           2    defaults
requests                  2.11.1                   py35_0    defaults
ruamel_yaml               0.11.14                  py35_0    defaults
scipy                     0.17.1                    <pip>
scipy                     0.18.1          np111py35_nomkl_0  [nomkl]  defaults
setuptools                27.2.0                   py35_0    defaults
simplegeneric             0.8.1                    py35_1    defaults
six                       1.10.0                   py35_0    defaults
sqlite                    3.13.0                        0    defaults
terminado                 0.6                      py35_0    defaults
tk                        8.5.18                        0    defaults
tornado                   4.4.2                    py35_0    defaults
traitlets                 4.3.1                    py35_0    defaults
wcwidth                   0.1.7                    py35_0    defaults
xz                        5.2.2                         0    defaults
yaml                      0.1.6                         0    defaults
zlib                      1.2.8                         3    defaults
```

**Comment by ilanschnell** [comment]
Ah, I see the problem.  The old `.egg-info` files are being read, which is also why conda lists them as `<pip>`.  I'm not sure what to do about that right now.  Is this issue blocking you?

**Comment by dsludwig** [comment]
Shouldn't those files be cleaned up by the upgrade process? I would expect the files in the old `conda-meta/*.json` files to be unlinked/deleted.

**Comment by ilanschnell** [comment]
Currently only the old `conda-meta/*.json` files are being cleaned up.  The problem is that the packages listed as `<pip>` don't have a `conda-meta/*.json` file.  The best solution would be to remove not only the old `conda-meta/*.json` files, but also the files they list *unless* the file was part of a package which was just installed.

**Comment by dsludwig** [comment]
> The best solution would be to remove not only the old conda-meta/*.json files, but also the files they list unless the file was part of a package which was just installed.

Yes, that sounds like the right solution.

**Comment by dsludwig** [comment]
> The problem is that the packages listed as <pip> don't have a `conda-meta/*.json` file

Is this before or after the `conda-meta` files are cleaned up? Would they not show up if the files listed in the `json` file were deleted before removing the `conda-meta` entry?

**Comment by ilanschnell** [comment]
@dsludwig I've added the implementation of removing the unused files listed in `conda-meta/*.json` in #48.  The files are removed right before the corresponding conda-meta file itself is removed.

**Comment by dsludwig** [comment]
That works perfectly! Thanks @ilanschnell

**Comment by ilanschnell** [comment]
#46 is now merged.  I also fixed the typo in the help text, and released version 1.5.0 which includes the new `-u` option.
